### ISSUE TITLE
Webhook doesn't work on API inserted row

### ISSUE THREAD

**Issue by ashkaptsov** [nocodb/nocodb#1440]
Hey all,

I have automatically new rows added to my nocoDB and would like to set up the webhook on new record. 
Webhook works when I insert the record manually, but doesn't work when I insert them via REST API

Is it a bug or there is other way to do it?
Thanx

**Comment by wingkwong** [comment]
Please share your project info.

**Comment by ashkaptsov** [comment]
Node: **v12.22.10**
Arch: **x64**
Platform: **linux**
Docker: **true**
Database: **pg**
ProjectOnRootDB: **true**
RootDB: **pg**
PackageVersion: **0.84.15**

**Comment by ashkaptsov** [comment]
any ideas?

**Comment by wingkwong** [comment]
Is there any error message you can share?

**Comment by ashkaptsov** [comment]
nope, just webhook doesn't trigger on insert if it is inserted by REST API, If I insert manually a row webhook triggers.

**Comment by wingkwong** [comment]
Just tested it. I'm able to trigger the webhook via REST API. Could you please share more details? Like if you've specified extra conditions?

Here's my testing result:

- I inserted `1` manually in GUI 
- Webhook triggered
- I used REST API to insert `2`
- Webhook triggered

<img width="638" alt="image" src="https://user-images.githubusercontent.com/35857179/158375478-66c6f4e7-00df-497e-8019-33ce8a733414.png">

Logs:

![image](https://user-images.githubusercontent.com/35857179/158375671-d77da652-0b15-4c6e-93d8-4ade2a97cbd7.png)

**Comment by wingkwong** [comment]
The first screenshot is from [Audit](https://docs.nocodb.com/setup-and-usages/audit). The second screenshot is from server log. If you don't have access to the logs, you can just go to Audit in GUI to see if the webhook is triggered or not.

Also, if possible, please share your webhook settings.

**Comment by wingkwong** [comment]
Please share your webhook settings. Also were 436 & 437 inserted by API?

**Comment by ashkaptsov** [comment]
Thats webhook settings:

<img width="1257" alt="Schermata 2022-03-16 alle 10 12 26" src="https://user-images.githubusercontent.com/70571543/158556250-4e403873-d91b-4276-a5cd-74bd7d8d37fc.png">

Audit:
I don't know why but there are no records in audit
<img width="940" alt="Schermata 2022-03-16 alle 10 13 49" src="https://user-images.githubusercontent.com/70571543/158556535-f8907735-c6b5-4b11-ad6d-8d58363d0f00.png">

records:

<img width="1122" alt="Schermata 2022-03-16 alle 10 15 31" src="https://user-images.githubusercontent.com/70571543/158556776-60cb712e-d3e9-4011-8bfc-e6e64cec6800.png">

**Comment by wingkwong** [comment]
Please DM me in Discord.

**Comment by wingkwong** [comment]
UPD: Only able to reproduce using user's environment. Suggested to check the outbound settings & logs in CapRover.

**Comment by o1lab** [comment]
Feel to reopen with more details. Problem might be has to do with server settings in caprover.

**Comment by HSUCHING** [comment]
I have also this issue.
And,it is ok when I test by this discord webhook in postman posting request,
it is ok when I test by slack webook in Nocodb dashboard.
So I guess there are some issues on discord app in Nocodb.

**Comment by onejgordon** [comment]
Apologies, our tests were using the batch API which did not have a webhook configured. Even without `xc-gui` webhook is behaving as expected.

**Comment by MattCarneiro** [comment]
> EDIT: Working as expected now. We were testing the wrong API call.
> 
> 
> 
> I'm experiencing this issue as well. Webhooks configured for "after update" and "after insert" do not fire when creating or updating records via the REST API. Both work as designed when changes are made in the Noco UI.
> 
> 
> 
> *Project Details*
> 
> 
> 
> Host: Pikapods
> 
> NocoDB used as docker: false
> 
> NocoDB version: 0.109.3
> 
> Database used in NC_DB_URL: pg
> 
> Project was created by clicking: New Project



How did you solve this?

I am having the same issue

**Comment by ashkaptsov** [comment]
I just upgraded or reinstalled(don't remember anymore), then it worked

Il giorno lun 8 lug 2024 alle ore 20:55 Elementttt ***@***.***>
ha scritto:

**Comment by kulkreate** [comment]
Also experiencing this issue. Everything up to date.

**Comment by timothyjoh** [comment]
Also experiencing with the Self-hosted version. n8n will write data to a row, which should trigger the webhook back to n8n for a different workflow. Circular like this for 3 steps on the same row. Works when manually going into the GUI and editing a null field with a few spaces, it triggers the update webhook again.

I can work around this by chaining the n8n workflows together, but not ideal as data can be entered in any order, and partial data too. Needs to trigger one of 3 workflows based on a `status` column.

**Comment by that0n3guy** [comment]
I'm experiencing this as well but for and UPDATED row.   Here is my webook:

![Image](https://github.com/user-attachments/assets/dcd7873c-621e-40d1-bdd5-27b64aeb0018)

When I update a row via n8n the webhook does NOT trigger.   

Maybe this is an n8n integration issue?  I see a previous user had this issue on n8n as well.

**Also, I'm using https://app.nocodb.com**

**Comment by that0n3guy** [comment]
Quick update.   I got this to work, but I had to do 2 things:

# Thing 1: 
I had to set the conditions to "contains any of", like so: 

![Image](https://github.com/user-attachments/assets/9788e57b-25f9-4b84-8e10-7cc0ca739258)

Even though the field is a single select.

---
# Thing 2:
I had to use a HTTP call and NOT the n8n nocodb node.  The HTTP node works (see image below) NOT the.   

### This triggers webhook:

![Image](https://github.com/user-attachments/assets/56fdbca7-9839-40ab-8f87-26d474d7e96c | width=100)

![Image](https://github.com/user-attachments/assets/0fcdd6ce-7ce2-4806-a034-3483b7170dac)
![Image](https://github.com/user-attachments/assets/98607659-f963-4c3d-a047-0f26f99bb1d1)


---


## This does NOT trigger webhook
![Image](https://github.com/user-attachments/assets/e17d8be0-1949-4373-8b68-c003c3829f26)

![Image](https://github.com/user-attachments/assets/a8ea134e-04ec-4241-a265-85c4e2372608)

**Comment by that0n3guy** [comment]
@o1lab With me using nocodb cloud and having the example above.   Can this get re-opened?

Thanks!

**Comment by that0n3guy** [comment]
Well, I have an example where I can't get this to work at all with insert (my previous screenshots were for "update".

Again, using nocodb.com

![Image](https://github.com/user-attachments/assets/42d03581-2971-4938-9d2d-c7355b333263)

**Comment by that0n3guy** [comment]
Ok, I may have figured this out.   

Sending a body like this does NOT trigger webhooks:

```
[
  {
    "url": "{{ $json.url }}",
    "stage": "new"
  }
]
```

Sending a body like this DOES trigger webhooks:
```
{
  "url": "{{ $json.url }}",
  "stage": "new"
}
```


Both of these will create the row just fine. 

This is the case for both UPDATE and INSERT. 

Update example:

DOES trigger webhooks: 
`{ "url": "{{ $('If URL').first().json.url }}", "id": {{ $json.id }} } `

DOES NOT trigger webhooks:
`[{ "url": "{{ $('If URL').first().json.url }}", "id": {{ $json.id }} } ]`

**Comment by dstala** [comment]
The array format `[ ]` in the payload causes the webhook handler to interpret it as a trigger for bulk insert or bulk update operations, even though the array size was one. We are currently revamping our webhook triggers by consolidating bulk triggers with their corresponding individual triggers, which should address this behaviour.

**Comment by that0n3guy** [comment]
@dstala I'm assuming the n8n is using the array format by default which would be why it doesn't work then.    

Thanks!

**Comment by dstala** [comment]
@ribtoks please add more details about your scenario. note that
- webhook trigger for bulk operations are separate
- webhook with conditions are triggered only when the condition is met for the first time https://docs.nocodb.com/automation/webhook/create-webhook/#webhook-with-conditions
